File: registration_system_rs/src/challenge/domain/challenge.rs
use crate::activity::domain::Activity;
use chrono::NaiveDateTime;
use rust_decimal::Decimal;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChallengeStatus {
    Open,
    Matched,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChallengeKind {
    Team,
    Individual,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ChallengePaymentMode {
    Prepaid,
    Postpaid,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum IndividualAcceptancePaymentStatus {
    Unpaid,
    Paid,
    Cancelled,
}

impl ChallengeKind {
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Team => "team",
            Self::Individual => "individual",
        }
    }

    pub fn from_db_str(value: &str) -> Self {
        match value {
            "individual" => Self::Individual,
            _ => Self::Team,
        }
    }
}

impl ChallengePaymentMode {
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Prepaid => "prepaid",
            Self::Postpaid => "postpaid",
        }
    }

    pub fn from_db_str(value: &str) -> Self {
        match value {
            "prepaid" => Self::Prepaid,
            _ => Self::Postpaid,
        }
    }
}

impl IndividualAcceptancePaymentStatus {
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Unpaid => "unpaid",
            Self::Paid => "paid",
            Self::Cancelled => "cancelled",
        }
    }

    pub fn from_db_str(value: &str) -> Self {
        match value {
            "paid" => Self::Paid,
            "cancelled" => Self::Cancelled,
            _ => Self::Unpaid,
        }
    }
}

impl ChallengeStatus {
    pub fn as_db_str(self) -> &'static str {
        match self {
            Self::Open => "open",
            Self::Matched => "matched",
            Self::Cancelled => "cancelled",
        }
    }

    pub fn from_db_str(value: &str) -> Self {
        match value {
            "matched" => Self::Matched,
            "cancelled" => Self::Cancelled,
            _ => Self::Open,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Challenge {
    pub id: String,
    pub title: String,
    pub kind: ChallengeKind,
    pub payment_mode: ChallengePaymentMode,
    pub host_team_id: Option<i64>,
    pub host_user_id: i64,
    pub guest_team_id: Option<i64>,
    pub accepted_by_user_id: Option<i64>,
    pub activity_id: Option<String>,
    pub holding_date: NaiveDateTime,
    pub start_time: NaiveDateTime,
    pub end_time: NaiveDateTime,
    pub location: String,
    pub location_latitude: Option<f64>,
    pub location_longitude: Option<f64>,
    pub players_per_team: i32,
    pub min_players: Option<i32>,
    pub max_players: Option<i32>,
    pub fee_per_person: Option<Decimal>,
    pub note: Option<String>,
    pub status: ChallengeStatus,
    pub accepted_at: Option<NaiveDateTime>,
    pub cancelled_at: Option<NaiveDateTime>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

impl Challenge {
    pub fn signup_capacity(&self) -> i32 {
        self.max_signup_players()
    }

    pub fn min_signup_players(&self) -> i32 {
        match self.kind {
            ChallengeKind::Individual => self.min_players.unwrap_or(self.players_per_team * 2),
            ChallengeKind::Team => self.players_per_team,
        }
    }

    pub fn max_signup_players(&self) -> i32 {
        match self.kind {
            ChallengeKind::Individual => self.max_players.unwrap_or(self.players_per_team * 2 + 4),
            ChallengeKind::Team => self.players_per_team,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ChallengeSummary {
    pub challenge: Challenge,
    pub host_team_name: String,
    pub host_team_credit_score: i32,
    pub host_team_trust_label: String,
    pub guest_team_name: Option<String>,
    pub guest_team_credit_score: Option<i32>,
    pub guest_team_trust_label: Option<String>,
    pub current_team_relation: Option<String>,
    pub accepted_count: i32,
    pub current_user_joined: bool,
    pub can_accept: bool,
    pub individual_participant_preview: Vec<ChallengeIndividualParticipant>,
}

#[derive(Debug, Clone)]
pub struct ChallengeIndividualParticipant {
    pub user_id: i64,
    pub display_name: String,
    pub avatar_url: Option<String>,
}

#[derive(Debug, Clone)]
pub struct CurrentUserIndividualAcceptance {
    pub payment_status: IndividualAcceptancePaymentStatus,
    pub payment_deadline_at: Option<NaiveDateTime>,
    pub payment_order_no: Option<String>,
}

#[derive(Debug, Clone)]
pub struct ChallengeDetail {
    pub summary: ChallengeSummary,
    pub activity: Option<Activity>,
    pub individual_participants: Vec<ChallengeIndividualParticipant>,
    pub current_user_acceptance: Option<CurrentUserIndividualAcceptance>,
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;

    fn challenge(kind: ChallengeKind, players_per_team: i32) -> Challenge {
        let now = Utc::now().naive_utc();
        Challenge {
            id: "challenge-a".to_string(),
            title: "测试约队".to_string(),
            kind,
            payment_mode: ChallengePaymentMode::Postpaid,
            host_team_id: Some(1),
            host_user_id: 1,
            guest_team_id: None,
            accepted_by_user_id: None,
            activity_id: None,
            holding_date: now,
            start_time: now,
            end_time: now,
            location: "测试球场".to_string(),
            location_latitude: None,
            location_longitude: None,
            players_per_team,
            min_players: None,
            max_players: None,
            fee_per_person: None,
            note: None,
            status: ChallengeStatus::Open,
            accepted_at: None,
            cancelled_at: None,
            created_at: now,
            updated_at: now,
        }
    }

    #[test]
    fn individual_challenge_defaults_min_players_to_both_sides() {
        assert_eq!(
            challenge(ChallengeKind::Individual, 8).min_signup_players(),
            16
        );
        assert_eq!(
            challenge(ChallengeKind::Individual, 5).min_signup_players(),
            10
        );
    }

    #[test]
    fn individual_challenge_defaults_max_players_to_both_sides_plus_buffer() {
        assert_eq!(
            challenge(ChallengeKind::Individual, 8).max_signup_players(),
            20
        );
        assert_eq!(
            challenge(ChallengeKind::Individual, 5).max_signup_players(),
            14
        );
    }

    #[test]
    fn individual_challenge_can_override_signup_limits() {
        let mut challenge = challenge(ChallengeKind::Individual, 8);
        challenge.min_players = Some(10);
        challenge.max_players = Some(14);

        assert_eq!(challenge.min_signup_players(), 10);
        assert_eq!(challenge.max_signup_players(), 14);
    }

    #[test]
    fn team_challenge_capacity_uses_opponent_side_only() {
        assert_eq!(challenge(ChallengeKind::Team, 8).signup_capacity(), 8);
        assert_eq!(challenge(ChallengeKind::Team, 5).signup_capacity(), 5);
        assert_eq!(challenge(ChallengeKind::Team, 8).min_signup_players(), 8);
        assert_eq!(challenge(ChallengeKind::Team, 8).max_signup_players(), 8);
    }
}

File: registration_system_backend_fe/src/services/challenge.ts
import { http } from '@/utils/request'

export type ChallengeStatus = 'open' | 'matched' | 'cancelled'
export type ChallengeKind = 'team' | 'individual'

export interface Challenge {
  id: string
  title: string
  kind: ChallengeKind
  host_team_id: number | null
  host_user_id: number
  guest_team_id: number | null
  accepted_by_user_id: number | null
  activity_id: string | null
  holding_date: string
  start_time: string
  end_time: string
  location: string
  location_latitude: number | null
  location_longitude: number | null
  players_per_team: number
  fee_per_person: string | null
  note: string | null
  status: ChallengeStatus
  accepted_at: string | null
  cancelled_at: string | null
  created_at: string
  updated_at: string
}

export interface ChallengeSummary {
  challenge: Challenge
  host_team_name: string
  host_team_credit_score: number
  host_team_trust_label: string
  guest_team_name: string | null
  guest_team_credit_score: number | null
  guest_team_trust_label: string | null
  current_team_relation: string | null
  accepted_count: number
  current_user_joined: boolean
  can_accept: boolean
  individual_participant_preview: ChallengeIndividualParticipant[]
}

export interface ChallengeActivityRef {
  id: string
  name: string
  holding_date: string
  start_time: string
  end_time: string
  location: string
  home_team_id: number | null
  away_team_id: number | null
  players_per_team: number | null
}

export interface ChallengeIndividualParticipant {
  user_id: number
  display_name: string
  avatar_url: string | null
}

export interface ChallengeDetail {
  summary: ChallengeSummary
  activity: ChallengeActivityRef | null
  individual_participants: ChallengeIndividualParticipant[]
}

export interface UpdateChallengePayload {
  title: string
  holding_date: string
  start_time: string
  end_time: string
  location: string
  location_latitude?: number | null
  location_longitude?: number | null
  players_per_team: number
  fee_per_person?: string | null
  note?: string | null
}

export interface CreateIndividualChallengePayload extends UpdateChallengePayload {
  kind: 'individual'
  host_user_id: number
  host_team_id?: null
}

export const adminListChallenges = (params: {
  team_id?: number
  keyword?: string
  status?: ChallengeStatus
  kind?: ChallengeKind
  include_closed?: boolean
  limit?: number
  sort?: 'holding_date_asc' | 'holding_date_desc' | 'created_at_desc' | 'credit_desc'
}) =>
  http.get<ChallengeSummary[]>('/challenges', {
    params: {
      team_id: params.team_id,
      keyword: params.keyword,
      status: params.status,
      kind: params.kind,
      include_closed: params.include_closed ?? false,
      limit: params.limit ?? 50,
      sort: params.sort ?? 'holding_date_asc',
    },
  })

export const getAdminChallengeDetail = (challengeId: string) =>
  http.get<ChallengeDetail>(`/challenges/${challengeId}`)

export const createAdminChallenge = (payload: CreateIndividualChallengePayload) =>
  http.post<Challenge>('/challenges', payload)

export const updateAdminChallenge = (challengeId: string, payload: UpdateChallengePayload) =>
  http.patch<Challenge>(`/challenges/${challengeId}`, payload)

export const cancelAdminChallenge = (challengeId: string) =>
  http.post<Challenge>(`/challenges/${challengeId}/cancel`)
